backend/src/main/java/com/empire/Region.java

```diff
@@ -33,169 +37,223 @@
+			mods += Constants.perIdeologyTapestryRecruitmentModGlobal * numUniqueIdeologies(kingdom, w);
 		}
-		if (Ideology.RJINKU == NationData.getStateReligion(kingdom, w) && rulerBattled) mods += .5;
-		if (Ideology.TAPESTRY_OF_PEOPLE == NationData.getStateReligion(kingdom, w)) mods += .03 * numUniqueIdeologies(kingdom, w);
-		if (NationData.getStateReligion(kingdom, w).religion == Religion.IRUHAN && Ideology.TAPESTRY_OF_PEOPLE  == w.getDominantIruhanIdeology() && NationData.getStateReligion(kingdom, w).religion == Religion.IRUHAN) mods += .03 * numUniqueIdeologies(kingdom, w);
+
+		if (largestInRegion != null && !NationData.isFriendly(kingdom, largestInRegion.kingdom, w) && largestInRegion.hasTag(Constants.armyPillagersTag)) mods += Constants.armyPillagersRecruitmentMod;
+
 		return Math.max(0, base * mods);
 	}
 
+	// TODO: this belongs alongside the game constants, should determine a way to parameterize these function-type rules
+	public double calcSigningBonusMod(double signingBonus){
+		/* Original:
+		if (signingBonus == -1) return -0.5;
+		else if (signingBonus == -2) return  -1;
+		else if (signingBonus >= 1) return (Math.log(signingBonus) / Math.log(2)) * .5 + .5;
+		*/
+		return signingBonus <= 0 ? signingBonus * 0.5 : (Math.log(signingBonus) / Math.log(2)) * 0.5 + 0.5;
+	}
+
+	public double calcTaxIncome(World w, List<Character> governors, double taxRate, double rationing) {
+		double base = population * Constants.taxPerPop;
+		double unrest = calcUnrest(w);
+		if (unrest > Constants.unrestTaxEffectThresh) base *= 1.0 - (unrest - Constants.unrestTaxEffectThresh);
 
-	public double calcTaxIncome(World w, ArrayList<Character> governors, double taxRate, double rationing) {
-		double base = population / 10000.0;
 		double mods = taxRate;
-		double unrest = calcUnrest(w);
-		NationData wKingdom = w.getNation(kingdom);
-		if (unrest > .25) base *= 1.25 - unrest;
-		if (noble != null && noble.hasTag("Frugal")) mods += .5;
-		if (noble != null && noble.hasTag("Hoarding")) mods -= .35;
+
 		if (governors != null) {
 			for (Character c : governors) {
-				mods += c.calcLevel("governor") * .5 + 1;
+				mods += c.calcLevel(Constants.charDimGovernor) * Constants.perLevelGovernTaxMod + 1;
 			}
 		}
-		if (wKingdom.hasTag("Coast-Dwelling") && isCoastal(w)) mods += .12;
-		if (wKingdom.hasTag("Mercantile")) mods += .15;
+
+		if (noble != null && noble.hasTag(Constants.nobleFrugalTag)) mods += Constants.nobleFrugalMod;
+		if (noble != null && noble.hasTag(Constants.nobleHoardingTag)) mods += Constants.nobleHoardingMod;
+
+		NationData wKingdom = w.getNation(kingdom);
+		if (wKingdom.hasTag(Constants.nationCoastDwellingTag) && isCoastal(w)) mods += Constants.coastDwellingTaxMod;
+		if (wKingdom.hasTag(Constants.nationMercantileTag)) mods += Constants.mercantileTaxMod;
+		if (wKingdom.hasTag(Constants.nationWarlikeTag) && wKingdom.coreRegions.contains(w.regions.indexOf(this))) {
+			int conquests = 0;
```

Comment:
I'd very much like to break down every long function into smaller functional components.  But wouldn't want to do that without also creating the tests alongside it (this is one didn't get tests this go around).  It will get them next time though.